### PR TITLE
Reorganize code in preparation for scamper2 datatype

### PR DESCRIPTION
**Pull request by SaiedKazemi**
This commit does not change the functionality of traceroute-caller.
It only reorganizes its code so that regular traceroutes can be easily
added in a separate commit.

Changes were tested locally via go test and also docker-compose.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/134)

### CODE REVIEW

**Review by cristinaleonr — COMMENTED** [review]
Looks like there's a 4.4% test coverage decrease. 

Some of the new functions that are missing coverage are `parser.New()` and `tracer.Validate()`. It would be great if we can add some tests for these.

> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-:-Mqp3Qf6AAL-xyfh4HKl:b-n69kim)* status: 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[caller.go, line 90 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp4U8b2LRl9o1I3Aku:-Mqp4U8b2LRl9o1I3Akv:bbm4uz9) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/caller.go#L90)):*
> ```Go
> 	}
> 	if err := scamper.Validate(); err != nil {
> 		logFatal(err)
> ```

As discussed, it would be great if we could add a bit more context to these errors, similar to line 113.
___
*[parser/parser.go, line 30 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp47lk9x1Jvq-x2u3f:-Mqp47lk9x1Jvq-x2u3g:btgpj88) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/parser/parser.go#L30)):*
> ```Go
> }
> 
> // CyclestartLine contains the information about the scamper "cyclestart"
> ```

Nitpick: missing period.
___
*[parser/parser.go, line 40 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp4mhJDcHVRfrLXzt6:-Mqp4mhJDcHVRfrLXzt7:b-iqqez) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/parser/parser.go#L40)):*
> ```Go
> 
> // CyclestopLine contains the ending details from the scamper tool.  ID,
> // ListName, hostname seem to match CyclestartLine
> ```

Same as above. Also, looks like`Hostname` should be capitalized.

**Review by stephen-soltesz — CHANGES_REQUESTED** [review]
> Reviewed 9 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/134)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[caller.go, line 33 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp6gaP4Xobbdg3Op7Y:-Mqp6gaP4Xobbdg3Op7Z:b64qqop) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/caller.go#L33)):*
> ```Go
> 	scamperBin          = flag.String("scamper.bin", "scamper", "The path to the scamper binary.")
> 	scamperTimeout      = flag.Duration("scamper.timeout", 900*time.Second, "How long to wait for scamper to complete a traceroute.")
> 	scamperTraceType    = flag.String("scamper.trace-type", "mda", "Specify the type of traceroute to run (currently only mda).")
> ```

In context, if this were "tracelb" it would be obvious to a reader that the "*Tracelb*" variables are related to this mode. It's not clear how a reader would otherwise know to associate "mda" with these other options.

Also, please consider using the https://pkg.go.dev/github.com/m-lab/go@v0.1.45/flagx#Enum to limit options to pre-defined values.
___
*[caller.go, line 73 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp7XjD-v7tMspzE6PA:-Mqp7XjD-v7tMspzE6PB:b-9xktnk) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/caller.go#L73)):*
> ```Go
> 	// already exists before running actual traceroutes which can
> 	// happen concurrently.
> 	exec.Command(*scamperBin, "-I", "trace -P icmp-paris 127.0.0.1", "-o-", "-O", "json").Run()
> ```

This feels weird. Is the issue that the logic later may start two processes that both try to create this directory and one may fail?

I would rather see the directly created explicitly or from the Dockerfile before starting this process. `main` does not feel like the right place.
___
*[parser/parser.go, line 33 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp8m6M7qdugzudWIZA:-Mqp8m6M7qdugzudWIZB:b8qx7fu) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/parser/parser.go#L33)):*
> ```Go
> type CyclestartLine struct {
> 	Type      string  `json:"type"`
> 	ListName  string  `json:"list_name" bigquery:"list_name"`
> ```

Saied, You mentioned this a few times, but I thought it was in reference to the old parser schema. For a new schema, I would recommend treating all fields consistently for the schema.

Because the v2 pipeline writes JSONL to GCS, and BigQuery loads the data from there, I *think* we want to add `bigquery` tags everywhere that mirror the JSON tag name. Otherwise, when we create the schema using `update-schema` command line tool from m-lab/etl it will create the CamelCase field names for those without a bigquery tag, and the BigQuery load will fail to find all the data.. I *think*.
___
*[tracer/scamper.go, line 54 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp4AGhAzjFoXWdMmKK:-Mqp4AGiASsyy68ELrI9:bx6jp4l) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/tracer/scamper.go#L54)):*
> ```Go
> }
> 
> // CachedTrace creates a traceroute from the traceroute cache.
> ```

I didn't realize until I read the code that this writes a file.
___
*[tracer/scamper.go, line 111 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp5JOCA4hW8eZ4e4BI:-Mqp5JOCA4hW8eZ4e4BJ:b9nwghh) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/tracer/scamper.go#L111)):*
> ```Go
> 	// fail because of them.
> 	var cmd []string
> 	if s.noArgs {
> ```

If this were me, rather than adding additional logic for the function under test, I might create a super simple shell script that "does the right thing" if "echo, yes, and false` all do the wrong thing. Then this condition is not needed.

**Comment by SaiedKazemi** [comment]
___
*[caller.go, line 33 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp6gaP4Xobbdg3Op7Y:-Mr2Zx6k97GQCh67U3Xq:b-kt32za) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/caller.go#L33)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

In context, if this were "tracelb" it would be obvious to a reader that the "*Tracelb*" variables are related to this mode. It's not clear how a reader would otherwise know to associate "mda" with these other options.

Also, please consider using the https://pkg.go.dev/github.com/m-lab/go@v0.1.45/flagx#Enum to limit options to pre-defined values.
</blockquote></details>

Changed the message to show "tracelb" corresponds to the "mda" traceroute type.

Used flagx's Enum.

**Comment by SaiedKazemi** [comment]
___
*[caller.go, line 73 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp7XjD-v7tMspzE6PA:-Mr2_Jcy0UrUIDw_HtAi:b78kf5o) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/caller.go#L73)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

This feels weird. Is the issue that the logic later may start two processes that both try to create this directory and one may fail?

I would rather see the directly created explicitly or from the Dockerfile before starting this process. `main` does not feel like the right place.
</blockquote></details>

I agree that this is an environment issue that doesn't belong to traceroute-caller.  However, because traceroute-caller can be invoked directly, Dockerfile won't help.  That's why I put the workaround here until scamper is fixed but I have removed it now.

**Comment by SaiedKazemi** [comment]
___
*[caller.go, line 90 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp4U8b2LRl9o1I3Aku:-Mr2aIaO-zn2V53xl0EX:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/caller.go#L90)):*
<details><summary><i>Previously, cristinaleonr (Cristina Leon) wrote…</i></summary><blockquote>

As discussed, it would be great if we could add a bit more context to these errors, similar to line 113.
</blockquote></details>

Done.

**Comment by SaiedKazemi** [comment]
___
*[parser/parser.go, line 30 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp47lk9x1Jvq-x2u3f:-Mr2aMPT15k0KI9WhrYM:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/parser/parser.go#L30)):*
<details><summary><i>Previously, cristinaleonr (Cristina Leon) wrote…</i></summary><blockquote>

Nitpick: missing period.
</blockquote></details>

Done.

**Comment by SaiedKazemi** [comment]
___
*[parser/parser.go, line 33 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp8m6M7qdugzudWIZA:-Mr2aPv60Ygt39-MdVAS:b-934o3w) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/parser/parser.go#L33)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Saied, You mentioned this a few times, but I thought it was in reference to the old parser schema. For a new schema, I would recommend treating all fields consistently for the schema.

Because the v2 pipeline writes JSONL to GCS, and BigQuery loads the data from there, I *think* we want to add `bigquery` tags everywhere that mirror the JSON tag name. Otherwise, when we create the schema using `update-schema` command line tool from m-lab/etl it will create the CamelCase field names for those without a bigquery tag, and the BigQuery load will fail to find all the data.. I *think*.
</blockquote></details>

Struct fields tags in parser.go now include bigquery:"xxx".

**Comment by SaiedKazemi** [comment]
___
*[parser/parser.go, line 40 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp4mhJDcHVRfrLXzt6:-Mr2aiHO00Im0eSZ6gL4:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/parser/parser.go#L40)):*
<details><summary><i>Previously, cristinaleonr (Cristina Leon) wrote…</i></summary><blockquote>

Same as above. Also, looks like`Hostname` should be capitalized.
</blockquote></details>

Done.

**Comment by SaiedKazemi** [comment]
___
*[tracer/scamper.go, line 54 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp4AGhAzjFoXWdMmKK:-Mr2alfA7u6fzt1reD9X:b-vtfl5v) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/tracer/scamper.go#L54)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I didn't realize until I read the code that this writes a file.
</blockquote></details>

The comment was from before.  I have now updated it to explicitly mention that it writes a file.

**Comment by SaiedKazemi** [comment]
___
*[tracer/scamper.go, line 111 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mqp5JOCA4hW8eZ4e4BI:-Mr2b2M6EFARzbYKm-tG:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/56166621cfa4f155188cc0d0a7c10d81205b16ee/tracer/scamper.go#L111)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

If this were me, rather than adding additional logic for the function under test, I might create a super simple shell script that "does the right thing" if "echo, yes, and false` all do the wrong thing. Then this condition is not needed.
</blockquote></details>

Done.

**Comment by SaiedKazemi** [comment]
I have added extra tests for higher coverage.  The only lines that are not covered now are the lines inside if conditions for errors that cannot be easily forced via package testing code.

ok  	github.com/m-lab/traceroute-caller	1.058s	coverage: 87.5% of statements
ok  	github.com/m-lab/traceroute-caller/hopannotation	0.646s	coverage: 93.6% of statements
ok  	github.com/m-lab/traceroute-caller/ipcache	3.020s	coverage: 100.0% of statements
ok  	github.com/m-lab/traceroute-caller/parser	0.034s	coverage: 100.0% of statements
ok  	github.com/m-lab/traceroute-caller/tracer	1.046s	coverage: 98.9% of statements
ok  	github.com/m-lab/traceroute-caller/triggertrace	0.088s	coverage: 96.3% of statements

**Review by stephen-soltesz — CHANGES_REQUESTED** [review]
> Reviewed 7 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/134)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @cristinaleonr, @gfr10598, and @SaiedKazemi)

___
*[tracer/scamper.go, line 34 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr2gl5I2AZUe0Ni2yE6:-Mr2gl5I2AZUe0Ni2yE7:b-obzhsx) ([raw file](https://github.com/m-lab/traceroute-caller/blob/088ce2530282c998734b060d8154ff7c92152545/tracer/scamper.go#L34)):*
> ```Go
> 	if fileInfo, err := os.Stat(s.Binary); err != nil {
> 		return fmt.Errorf("failed to stat scamper binary (error: %v)", err)
> 	} else if fileInfo.IsDir() {
> ```

Please add a unit test where `s.Binary` is a plain file that is not executable.
___
*[tracer/scamper.go, line 39 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr2hT5N1Cy9Vl1PAgtK:-Mr2hT5N1Cy9Vl1PAgtL:b-vyrmx3) ([raw file](https://github.com/m-lab/traceroute-caller/blob/088ce2530282c998734b060d8154ff7c92152545/tracer/scamper.go#L39)):*
> ```Go
> 
> 	switch s.TraceType {
> 	case "MDA":
> ```

What is the code path that makes this option possible? Why do we want to support this as valid input?

**Comment by SaiedKazemi** [comment]
___
*[tracer/scamper.go, line 34 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr2gl5I2AZUe0Ni2yE6:-Mr3-Gk14qarM47wNKmS:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/088ce2530282c998734b060d8154ff7c92152545/tracer/scamper.go#L34)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Please add a unit test where `s.Binary` is a plain file that is not executable.
</blockquote></details>

Done.

**Comment by SaiedKazemi** [comment]
___
*[tracer/scamper.go, line 39 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr2hT5N1Cy9Vl1PAgtK:-Mr3-IKs3HigCHBsorAh:bdm0e61) ([raw file](https://github.com/m-lab/traceroute-caller/blob/088ce2530282c998734b060d8154ff7c92152545/tracer/scamper.go#L39)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

What is the code path that makes this option possible? Why do we want to support this as valid input?
</blockquote></details>

While TRC recognizes "mda" and "regular" in lower case as valid types, the tracer package can be used by other programs. too.  And because MDA is an acronym, the intention was for the package to allow both "mda" and "MDA".  But I removed the case for "MDA".

**Review by stephen-soltesz — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/134)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @cristinaleonr, @gfr10598, and @SaiedKazemi)

___
*[tracer/scamper.go, line 39 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr2hT5N1Cy9Vl1PAgtK:-Mr32wpi3W3BmeHd7waB:b-s6pka0) ([raw file](https://github.com/m-lab/traceroute-caller/blob/088ce2530282c998734b060d8154ff7c92152545/tracer/scamper.go#L39)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

While TRC recognizes "mda" and "regular" in lower case as valid types, the tracer package can be used by other programs. too.  And because MDA is an acronym, the intention was for the package to allow both "mda" and "MDA".  But I removed the case for "MDA".
</blockquote></details>

Thank you. Every line of code is something we potentially have to maintain and support. Asking callers to use a single abbreviation feels like a fair trade to me, since they won't maintain our code :)
___
*[tracer/scamper.go, line 36 at r3](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr32Y63-yooWFjcmMz0:-Mr32Y63-yooWFjcmMz1:b67ztep) ([raw file](https://github.com/m-lab/traceroute-caller/blob/30fa937e812d4d99ee911f22109f244c688d21e0/tracer/scamper.go#L36)):*
> ```Go
> 		return fmt.Errorf("failed to stat scamper binary (error: %v)", err)
> 	}
> 	if fileInfo.IsDir() {
> ```

What if this is a named pipe or a device file? 

There exists a simpler and complete condition.

**Comment by SaiedKazemi** [comment]
___
*[tracer/scamper.go, line 36 at r3](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr32Y63-yooWFjcmMz0:-Mr37XaDEwMMG13dvTI8:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/30fa937e812d4d99ee911f22109f244c688d21e0/tracer/scamper.go#L36)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

What if this is a named pipe or a device file? 

There exists a simpler and complete condition.
</blockquote></details>

Done.

**Review by stephen-soltesz — APPROVED** [review]
:lgtm: - do not merge until approval from @cristinaleonr 

> Reviewed 1 of 3 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-:-Mr3bdvw28XCCcSDqziN:bhoh4qs)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @cristinaleonr and @gfr10598)

___
*[tracer/scamper.go, line 40 at r4](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr3_xXE04US3LtY5kU0:-Mr3_xXE04US3LtY5kU1:b-d325lj) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c1ac2fcdae2093b32fff0dca55b00fd062c0cf1e/tracer/scamper.go#L40)):*
> ```Go
> 		return fmt.Errorf("scamper binary is not a regular file")
> 	}
> 	if fileMode&0100 == 0 {
> ```

You're clearly in my wheel house. :)

You don't have to solve this in general, but just to point out that this is more complex still.

What would this return for a non-owner user? e.g. binary is owned by root, but not executable by group or users, e.g. 0700.

This should not impact M-Lab's use-case b/c we run the containers as uid:0.

**Comment by SaiedKazemi** [comment]
___
*[tracer/scamper.go, line 40 at r4](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr3_xXE04US3LtY5kU0:-Mr45uKy9nX2C3YTLE2c:b2dszid) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c1ac2fcdae2093b32fff0dca55b00fd062c0cf1e/tracer/scamper.go#L40)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

You're clearly in my wheel house. :)

You don't have to solve this in general, but just to point out that this is more complex still.

What would this return for a non-owner user? e.g. binary is owned by root, but not executable by group or users, e.g. 0700.

This should not impact M-Lab's use-case b/c we run the containers as uid:0.
</blockquote></details>

Sorry Stephen, I am not sure if I understand.

If mode is 0700, all is OK because fileMore&0100 will be 1 and the if condition isn't true.  Isn't this what we want?

**Comment by SaiedKazemi** [comment]
Thanks for LGTM but I am not merging because, sorry, I don't understand your comment on fileMode&0100.

**Review by cristinaleonr — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-:-Mr4VMjABiOogEE3RF9n:bg7igkv)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @cristinaleonr)

**Review by stephen-soltesz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/134)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @cristinaleonr)

___
*[tracer/scamper.go, line 40 at r4](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr3_xXE04US3LtY5kU0:-Mr4Yu2mCH1PbnI4Zxqx:bhjy7uf) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c1ac2fcdae2093b32fff0dca55b00fd062c0cf1e/tracer/scamper.go#L40)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

Sorry Stephen, I am not sure if I understand.

If mode is 0700, all is OK because fileMore&0100 will be 1 and the if condition isn't true.  Isn't this what we want?
</blockquote></details>

Could non-root, non-owner user running this command execute a binary with mode 0700? (I read this as a false-positive for a `Validate` method).

**Comment by SaiedKazemi** [comment]
___
*[tracer/scamper.go, line 40 at r4](https://reviewable.io/reviews/m-lab/traceroute-caller/134#-Mr3_xXE04US3LtY5kU0:-MrMOOZn6Ht5ZZDkqgBA:b-4qiutq) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c1ac2fcdae2093b32fff0dca55b00fd062c0cf1e/tracer/scamper.go#L40)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Could non-root, non-owner user running this command execute a binary with mode 0700? (I read this as a false-positive for a `Validate` method).
</blockquote></details>

The non-root, non-owner user will get a permission denied error and, in that sense, you're right that `Validate()` can be seen as a false-positive method.

I should have commented that, in its current form, `Validate()` is meant to validate the traceroute type and, although it checks that the scamper binary exists,`Validate()` is not meant to be a full-fledged validator.  If some parameters are invalid (e.g., OutputPath), they will cause an error (which will be handled appropriately) elsewhere in the code.

In light of this feedback, I will add complete checks in a separate PR.

**Comment by SaiedKazemi** [comment]
Thank you both for your reviews.

**Pull request merged**
